File: Algorithm/Java/src/main/java/com/yugutou/charpter18_backtracking/level2/Subsets.java
package com.yugutou.charpter18_backtracking.level2;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;

class Subsets {
    // 存放符合条件结果的集合
    List<List<Integer>> result = new ArrayList<>();
    // 用来存放符合条件结果
    LinkedList<Integer> path = new LinkedList<>();

    public List<List<Integer>> subsets(int[] nums) {
        //空集合也是一个子集
        if (nums.length == 0) {
            result.add(new ArrayList<>());
            return result;
        }
        subsetsHelper(nums, 0);
        return result;
    }

    private void subsetsHelper(int[] nums, int startIndex) {
//「遍历这个树的时候，把所有节点都记录下来，就是要求的子集集合」。
        result.add(new ArrayList<>(path));
        if (startIndex >= nums.length) {
            return;
        }
        for (int i = startIndex; i < nums.length; i++) {
            path.add(nums[i]);
            subsetsHelper(nums, i + 1);
            path.removeLast();
        }
    }

    public static void main(String[] args) {
        Subsets subsets = new Subsets();
        int[] nums = {1, 2, 3};
        System.out.println(subsets.subsets(nums));
    }
}

File: Algorithm/Leetcode/Code_Materials/c_or_cpp/charpter18_backtrace/level2/subsets.cpp
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

class Subsets {
    vector<vector<int>> result;
    queue<int> path;
    
public:

   void subsetsHelper(vector<int>& nums, int startIndex) {
         result.push_back(path);
        
        if (startIndex >= nums.size()) {
            return;
        }
        for (int i = startIndex; i < nums.size(); i++) {
            path.push(nums[i]);
            subsetsHelper(nums, i + 1);
            path.pop();
        }
    }

    vector<vector<int>> subsets(vector<int>& nums) {
        if (nums.empty()) {
            result.push_back(vector<int>());
            return result;
        }
        subsetsHelper(nums, 0);
        return result;
    }
    
 
};

int main() {
    int nums[] = {1, 2, 3};
    Subsets subsetter;
    vector<vector<int>> subsets = subsetter.subsets(nums);
    for (auto subset : subsets) {
        for (auto num : subset) {
            cout << num << " ";
        }
        cout << endl;
    }
    return 0;
}